R/preprocessing.R

```diff
@@ -0,0 +1,152 @@
+##'
+##' @description `r lifecycle::badge("experimental")`
+##'
+##' @param data Data frame with a `date` column. The other columns depend on the
+##'   model that the data are to be used, e.g. [estimate_infections()] or
+##'   [estimate_secondary()]. See the documentation there for the expected
+##'   format.
+##' @param missing Character. Options are "missing" (the default), "accumulate"
+##'   and "zero". This determines how missing dates in the data are interpreted.
+##'   If set to "missing", any missing dates in the observation data will be
+##'   interpreted as missing and skipped in the likelihood. If set to
+##'   "accumulate", modelled observations will be accumulated and added to the
+##'   next non-missing data point. This can be used to model incidence data that
+##'   is reported at less than daily intervals. If set to "accumulate", the
+##'   first data point is not included in the likelihood (unless `initial` is
+##'   set to a non-zero value) but used only to reset modelled observations to
+##'   zero. If "zero" then all observations on missing dates will be assumed to
+##'   be of value 0.
+##' @param na Character. How to process dates that have NA values. The options
+##'   available are the same ones as for the `missing` argument.
+##' @param initial_accumulate Integer. The number of initial dates to accumulate
+##'   if `missing` or a column name is set to `"accumulate"`. This number of
+##'   dates is added to the beginning of the data set to be accumulated onto the
+##'   first data point. This is useful, for example, for modelling weekly
+##'   incidence data, in which case this should be set to 7. If accumulating and
+##'   the first data point is one to accumulate and this is not set, then that
+##'   data point will be removed with a warning.
+##' @param obs_column Character. If given, only the column specified here will
+##'   be used for checking missingness. This is useful if using a data set that
+##'   has NA values in multiple columns, but only one of them corresponds to
+##'   observations that are to be processed here. By default, all columns will
+##'   be processed and, e.g. `accumulate` will be set if any column is NA in any
+##'   given row.
+##' @return a data.table with an `accumulate` column that indicates whether
+##'   values are accumulated (see the documentation of the `data` argument in
+##'   [estimate_infections()])
+##' @importFrom rlang arg_match
+##' @importFrom data.table as.data.table data.table
+##' @export
+##' @examples
+##' cases <- data.table::copy(example_confirmed)
+##' ## calculate weekly sum
+##' cases[, confirm := data.table::frollsum(confirm, 7)]
+##' ## limit to dates once a week
+##' cases <- cases[seq(7, nrow(cases), 7)]
+##' ## set the second observation to missing
+##' cases[2, confirm := NA]
+##' ## fill missing data
+##' fill_missing(cases, missing = "accumulate", initial_accumulate = 7)
+fill_missing <- function(data,
```

Comment:
This function errors if the data already has an `accumulate` column (from the assertion for column names). 

Should it rather throw a message and do nothing if it finds an `accumulate` column?

```r
library(data.table)
cases <- data.table::copy(example_confirmed)
## calculate weekly sum
cases[, confirm := data.table::frollsum(confirm, 7)]
## limit to dates once a week
cases <- cases[seq(7, nrow(cases), 7)]
## set the second observation to missing
cases[2, confirm := NA][, accumulate:=TRUE]
## fill missing data
fill_missing(cases, missing = "accumulate", initial_accumulate = 1, obs_column = "confirm")

Error in fill_missing(cases, missing = "accumulate", initial_accumulate = 1,  : 
  Assertion on 'names(data)' failed: Names must be disjunct from {'accumulate'}, but has elements {'accumulate'}.
```

Reply:
What would be the rationale for this? I don't really see when we'd want to fail silently if someone calls `fill_missing()` and we're not doing anything because an `accumulate` column already exists.

Reply:
Alternatively perhaps we could have a more informative error message?

Reply:
Alternatively perhaps we could have a more informative error message?

Reply:
I am pretty happy with it failing as `fill_missing` is optional. We could make it more informative but tbh I am happy

Reply:
> Alternatively perhaps we could have a more informative error message?

Yes, something like "data already has an "accumulate" column" is more friendly.

Reply:
I've decided against this as really this applies to all the `{checkmate}` checks and we don't want to change them all. I added some more clarifying documentation instead on the topic.